File: projeto_robo_c++/src/battery.cpp
#include "./include/battery.hpp"
#include "./include/navigation.hpp"
#include <fstream>

// Construtor para inicializar a bateria com capacidade máxima
Battery::Battery() : top(-1) {
    stack.resize(MAX_BATTERY, 1);
    top = MAX_BATTERY - 1;
}

// Verifica se a bateria está vazia
bool Battery::is_empty() const {
    return top == -1;
}

// Verifica se a bateria está cheia
bool Battery::is_full() const {
    return top == MAX_BATTERY - 1;
}

// Empilha (recarrega) uma unidade de bateria
void Battery::push(int value) {
    if (!is_full()) {
        stack[++top] = value;
    }
}

// Desempilha (consome) uma unidade de bateria
int Battery::pop() {
    if (!is_empty()) {
        return stack[top--];
    }
    return -1; // Indica que a bateria está vazia
}

// Obtém o nível atual da bateria
int Battery::battery_level() const {
    return top + 1;
}

// Função para simular recarregamento de bateria
void Battery::recharge_battery() {
    std::cout << "Recharging battery...\n";
    std::this_thread::sleep_for(std::chrono::seconds(RECHARGE_TIME));
    while (!is_full()) {
        push(1);
    }
    std::cout << "Battery fully recharged. Battery level: " << battery_level() << "\n";
}

// Gerencia a bateria enquanto move o robô
int Battery::manage_battery_on_move(int* actual, int* next_pos) {
    tMap mapa = read_map_from_file();
    if (mapa.map_data.empty()) {
        std::cerr << "Failed to read map data\n";
        return -1; // Indicando erro
    }

    int time = 0;
    int y_bat = next_pos[1];
    int x_bat = next_pos[0];

    if (y_bat < 0 || y_bat >= mapa.sz_y || x_bat < 0 || x_bat >= mapa.sz_x) {
        std::cerr << "Next position out of map bounds\n";
        return -1; // Indicando erro
    }

    int cell = mapa.map_data[8 - y_bat][x_bat];
    int level = battery_level();

    if (cell == 255 || cell == 191 || cell == 127 || cell == 63) { // Áreas navegáveis
        if (level > 1) {
            pop();
            actual[0] = x_bat;
            actual[1] = y_bat;
            std::cout << "Moved to (" << actual[0] << ", " << actual[1] << "), Battery: " << battery_level() << "\n";
            time += MOVE_TIME;
            std::cout << "2 seconds for move\n";
            std::this_thread::sleep_for(std::chrono::seconds(MOVE_TIME));

            // Verifica se a bateria está em 25% ou menos
            if (level - 1 < 2) {
                std::cout << "Battery low. Charging at (" << actual[0] << ", " << actual[1] << ")\n";
                recharge_battery();

                std::ofstream file("./data/charging.txt", std::ios::app);
                if (file) {
                    file << "Charge at (" << x_bat << ", " << y_bat << ")\n";
                } else {
                    std::cerr << "Failed to open charging log file\n";
                }
                time += RECHARGE_TIME;
            }
        } else {
            std::cout << "Not enough battery to move\n";
        }
    } else if (cell == 1) { // Área com impossibilidade de carga
        if (level > 2) {
            pop();
            actual[0] = x_bat;
            actual[1] = y_bat;
            std::cout << "Moved to (" << actual[0] << ", " << actual[1] << "), Battery: " << battery_level() << "\n";
            time += MOVE_TIME;
            std::cout << "2 seconds for move\n";
            std::this_thread::sleep_for(std::chrono::seconds(MOVE_TIME));
        } else {
            std::cout << "Battery too low to move to an area with impossibility of charge\n";
            // Recarrega a bateria antes de se mover
            std::cout << "Battery low. Charging at (" << actual[0] << ", " << actual[1] << ")\n";
            recharge_battery();
            pop();

            std::ofstream file("./data/charging.txt", std::ios::app);
            if (file) {
                file << "Charge at (" << actual[0] << ", " << actual[1] << ")\n";
            } else {
                std::cerr << "Failed to open charging log file\n";
            }
            time += RECHARGE_TIME;
        }
    }

    return time;
}

File: projeto_robo_python/src/robot_functions/battery.py
import time
from navigation import Map, Position

MAX_BATTERY = 4
RECHARGE_TIME = 4
MOVE_TIME = 2

class Battery:
    def __init__(self):
        self.stack = [1] * MAX_BATTERY
        self.top = MAX_BATTERY - 1

    def is_empty(self):
        return self.top == -1

    def is_full(self):
        return self.top == MAX_BATTERY - 1

    def push(self, value):
        if not self.is_full():
            self.top += 1
            self.stack[self.top] = value

    def pop(self):
        if not self.is_empty():
            value = self.stack[self.top]
            self.top -= 1
            return value
        return -1  # Indicates that the battery is empty

    def battery_level(self):
        return self.top + 1

    def recharge_battery(self):
        print("Recharging battery...")
        time.sleep(RECHARGE_TIME)  # Simulate 4 seconds of charging time
        while not self.is_full():
            self.push(1)
        print(f"Battery fully recharged. Battery level: {self.battery_level()}")

    def manage_battery_on_move(self, actual, next_pos):
        mapa = Map()
        mapa.read_map_from_file()
        if not mapa.map_data:
            print("Failed to read map data")
            return -1  # Indicating error

        time_spent = 0
        y_bat = next_pos[1]
        x_bat = next_pos[0]

        if y_bat < 0 or y_bat >= mapa.sz_y or x_bat < 0 or x_bat >= mapa.sz_x:
            print("Next position out of map bounds")
            return -1  # Indicating error

        cell = mapa.map_data[y_bat][x_bat]
        level = self.battery_level()

        if cell in [255, 191, 127, 63]:  # Navigable areas
            if level > 1:
                self.pop()
                actual[0], actual[1] = x_bat, y_bat
                time_spent += MOVE_TIME
                print("2 seconds for move")
                time.sleep(MOVE_TIME)
                print(f"Moved to ({actual[0]}, {actual[1]}), Battery: {self.battery_level()}")

                if level - 1 < 2:
                    print(f"Battery low. Charging at ({actual[0]}, {actual[1]})")
                    self.recharge_battery()
                    with open("./data/charging.txt", "a") as file:
                        file.write(f"Charge at ({x_bat}, {y_bat})\n")
                    time_spent += RECHARGE_TIME
            else:
                print("Not enough battery to move")
        elif cell == 1:  # Area with impossibility of charge
            if level > 2:
                self.pop()
                actual[0], actual[1] = x_bat, y_bat
                time_spent += MOVE_TIME
                print("2 seconds for move")
                time.sleep(MOVE_TIME)
                print(f"Moved to ({actual[0]}, {actual[1]}), Battery: {self.battery_level()}")
            else:
                print(f"Battery too low to move to an area with impossibility of charge. Charging at ({actual[0]}, {actual[1]})")
                self.recharge_battery()
                with open("./data/charging.txt", "a") as file:
                    file.write(f"Charge at ({actual[0]}, {actual[1]})\n")
                time_spent += RECHARGE_TIME
                self.pop()
                actual[0], actual[1] = x_bat, y_bat
                time_spent += MOVE_TIME
                print("2 seconds for move")
                time.sleep(MOVE_TIME)
                print(f"Moved to ({actual[0]}, {actual[1]}), Battery: {self.battery_level()}")

        return time_spent
